feat(slider): v1

bmdalex opened the pull request:
# feat(slider): v1

Closes #1539 

## Description

This PR consists of the first version of the `Slider` component. It incorporates:

### 1. Functionality (demo below):
![Screen Recording 2019-06-28 at 07 59 34](https://user-images.githubusercontent.com/5442794/60320962-a92f7e80-997b-11e9-954c-54e704c15a0a.gif)

### 2. Styling (according to design files and redlines for all themes):

| Theme Name | Slider flavor |
|:-------:|:------:|
| Base | ![Screenshot 2019-06-28 at 08 05 07](https://user-images.githubusercontent.com/5442794/60321064-f9a6dc00-997b-11e9-8f68-4299f4f5317d.png) |
| Teams | ![Screenshot 2019-06-28 at 07 58 09](https://user-images.githubusercontent.com/5442794/60321071-fdd2f980-997b-11e9-8ab8-cfde9ffb0a5c.png) 
| Teams Dark | ![Screenshot 2019-06-28 at 07 58 25](https://user-images.githubusercontent.com/5442794/60321078-03c8da80-997c-11e9-8361-9cd504896f38.png) |
| Teams HC | ![Screenshot 2019-06-28 at 07 58 44](https://user-images.githubusercontent.com/5442794/60321079-06c3cb00-997c-11e9-99b6-bfae38fbdd39.png) |

### 3. Accessibility:
  - leverages native keyboard navigation
  - adds correct accessibility attributes via the `sliderBehavior`

### 4. Doc examples and a few visual tests

## API

```typescript
export interface SliderProps {
  /**
   * Accessibility behavior if overridden by the user.
   * @default sliderBehavior
   */
  accessibility?: Accessibility

  /** The default value of the slider. */
  defaultValue?: SupportedIntrinsicInputProps['defaultValue'] // TODO 'value' type is React.ReactText but it doesn't work for 'defaultValue'

  /**
   * A slider can be read-only and unable to change states.
   * @default false
   */
  disabled?: SupportedIntrinsicInputProps['disabled']

  /**
   * A slider can take the width of its container.
   * @default false
   */
  fluid?: boolean

  /**
   * Callback that creates custom accessibility message a screen reader narrates when the value changes.
   * @param {SliderProps} props - Slider props.
   */
  getA11yValueMessageOnChange?: (props: SliderProps) => string

  /** Shorthand for the input component. */
  input?: ShorthandValue

  /** Ref for input DOM node. */
  inputRef?: React.Ref<HTMLElement>

  /**
   * The minimum value of the slider
   * @default 0
   */
  min?: SupportedIntrinsicInputProps['min']

  /**
   * The maximum value of the slider.
   * @default 100
   */
  max?: SupportedIntrinsicInputProps['max']

  /**
   * Called after item checked state is changed.
   * @param {SyntheticEvent} event - React's original SyntheticEvent.
   * @param {object} data - All props.
   */
  onChange?: ComponentEventHandler<SliderProps>

  /**
   * A number that specifies the granularity that the value must adhere to, or the special value 'any'.
   * A string value of any means that no stepping is implied, and any value is allowed
   * (barring other constraints, such as min and max).
   * @default 1
   */
  step?: SupportedIntrinsicInputProps['step']

  /** The value of the slider. */
  value?: React.ReactText

  /**
   * A slider can be displayed vertically.
   * @default false
   */
  vertical?: boolean
}
```

## TODO

- [x] Focus borders: still didn't find a way to add double borders to the slider's thumb
- [x] Remove slider marks in IE
- [x] Improve perf for recalculating props (accessibility props, styling props, etc) with every change of the slider's value
- [x] Add more visual tests

bmdalex committed:
feat(slider): v1

layershifter reviewed (commented):
Oops, CI fails and also there is warning somewhere is examples:
![image](https://user-images.githubusercontent.com/14183168/60442526-80c4b000-9c19-11e9-8191-d583118fa009.png)

layershifter commented on docs/src/examples/components/Slider/Types/SliderExample.shorthand.tsx:
```diff
@@ -0,0 +1,6 @@
+
+const SliderExample = () => <Slider />
+
+export default SliderExample
```
```suggestion
export default SliderExampleShorthand
```

bmdalex replied:
done thx

layershifter commented on docs/src/examples/components/Slider/Types/SliderExampleDisabled.shorthand.tsx:
```diff
@@ -0,0 +1,6 @@
+
+const SliderExampleDisabled = () => <Slider disabled />
+
+export default SliderExampleDisabled
```
```suggestion
export default SliderExampleDisabledShorthand
```

bmdalex replied:
done thx

layershifter commented on docs/src/examples/components/Slider/Types/index.tsx:
```diff
@@ -0,0 +1,21 @@
+      examplePath="components/Slider/Types/SliderExample"
+    />
+    <ComponentExample
+      title="Disabled"
```
`disabled` prop describes state

bmdalex replied:
done thx

layershifter commented on docs/src/examples/components/Slider/Variations/index.tsx:
```diff
@@ -0,0 +1,31 @@
+    <ComponentExample
+      title="Icon"
+      description="A slider can have an icon."
+      examplePath="components/Slider/Variations/SliderExampleIcon"
```
`icon` and `iconPosition` can be combined into single example, what do you think?

And also placed under `Slots` section

bmdalex replied:
done thx

bmdalex committed:
addressed comments and fixed build errors

bmdalex committed:
improved slider styles and functionality

codepretty commented on packages/react/src/components/Slider/Slider.tsx:
```diff
@@ -0,0 +1,236 @@
+            })}
+          </Ref>
+        </div>
+        {iconPosition !== 'start' && this.renderIcon(variables, styles)}
```
Should the icon optionally be a button instead of just icon since it needs the ability to perform some action when clicked? when clicked it should mute
![image](https://user-images.githubusercontent.com/828692/60930166-3cfd2680-a269-11e9-89c7-6fc174551124.png)

codepretty replied:
We will also need hover/focus styles for the icon button

bmdalex replied:
good point, let me think about it; will update the thread soon
@codepretty is the 20 volume mark part of the slider? I don't see that in the design files I have

bmdalex replied:
discussed with the team and we decided not to do that in v1 since it's not a requirement that's generic enough; we might reconsider; I will add an example on how to compose the `Slider` with `Button` component to achieve the behavior you're mentioning:
> it needs the ability to perform some action when clicked

codepretty replied:
The screenshot above is from the Windows default speaker controls - I was just trying to show a common volume control. Another requirement would be the icon updating at different points on the slider to indicate volume level. I'm not sure when something becomes a new component vs a usage example vs a prototype, but something to capture these requirements would be nice! Thank you @Bugaa92

bmdalex replied:
sure, np @codepretty 
I just added an example to show how to compose the Slider with Button component to achieve the behavior you're mentioning:

> it needs the ability to perform some action when clicked

so I'm resolving this comment

bmdalex committed:
- implemented border styles;
- fixed all styling issues in all browsers;
- implemented active styles and fixed existing bugs

bmdalex committed:
cleanup

bmdalex committed:
add screenshot tests

bmdalex committed:
example wip

bmdalex committed:
finished usage example

bmdalex committed:
removed icon and iconPosition

miroslavstastny reviewed (commented):
RTL is broken :-(
![slider-rtl](https://user-images.githubusercontent.com/9615899/61114613-fbe75b00-a490-11e9-8bca-17855046f37c.gif)

Both mouse and keyboard.

mnajdova commented on packages/react/src/components/Slider/Slider.tsx:
```diff
@@ -0,0 +1,276 @@
+import { SupportedIntrinsicInputProps } from '../../lib/htmlPropsUtils'
+import Box from '../Box/Box'
+
+const defaultPropValues = {
```
Why not inlining these to the defaultProps? It's hard to follow what is actually in the default props

bmdalex replied:
done, I was using them outside the component and thought it's more obvious, but managed to find a way, thanks!

mnajdova commented on packages/react/src/components/Slider/Slider.tsx:
```diff
@@ -0,0 +1,276 @@
+export interface SliderProps
+  extends UIComponentProps,
+    ChildrenComponentProps,
+    Omit<SupportedIntrinsicInputProps, 'defaultValue'> {
```
I understand that we want to reuse these handlers, but it seems a bit awkward why on some handlers, like `onChange` we provide the props, but on others, like `onMouseDown` we don't... Can we somehow unify this behavior?

bmdalex replied:
no idea how, but I don't think it's a blocker for this PR
I added the `onChange` here just because it's very important for the `Slider` (same is done in `Input`)...

mnajdova commented on packages/react/src/components/Slider/Slider.tsx:
```diff
@@ -0,0 +1,276 @@
+                value,
+                onChange: this.handleChange,
+                onFocus: this.handleFocus,
+                onMouseDown: this.handleMouseDown,
```
Shouldn't the handlers go in the `overrideProps`? This way those will be overriden if the user provides the same one.

bmdalex replied:
ooops... great catch, done

layershifter commented on packages/react/src/components/Slider/Slider.tsx:
```diff
@@ -0,0 +1,276 @@
+  step?: SupportedIntrinsicInputProps['step']
+
+  /** The value of the slider. */
+  value?: React.ReactText
```
This type looks strange... Can we use `number | string` instead?

bmdalex replied:
done, thanks

layershifter commented on packages/react/src/components/Slider/Slider.tsx:
```diff
@@ -0,0 +1,276 @@
+   * @param {SyntheticEvent} event - React's original SyntheticEvent.
+   * @param {object} data - All props.
+   */
+  onChange?: ComponentEventHandler<SliderProps>
```
```suggestion
  onChange?: ComponentEventHandler<SliderProps & { value: string }>
```

I believe that this is more accurate typing

mnajdova replied:
Basically for all components, the right type is `Props & State`. Would suggest addressing this for all components in a separate PR

bmdalex replied:
I don't agree with the `onChange?: ComponentEventHandler<SliderProps & { value: string }>` suggestion. `value` can be number as well. Also `value` is part of `SliderProps` so I don't see what the issue is exactly. Can I resolve?
@mnajdova @layershifter

mnajdova replied:
For this component this is definitely not a blocker, but it is something that we will need to tackle for other components.

layershifter replied:
Value in callback can't be `number`, please check it again.
https://github.com/stardust-ui/react/pull/1559#discussion_r302952170

bmdalex replied:
indeed, @layershifter is right; applied change but as @mnajdova says we need to align these interfaces in a separate PR; thanks guys, resolving 😃

layershifter commented on docs/src/examples/components/Slider/Playground.tsx:
```diff
@@ -0,0 +1,29 @@
+      step={step}
+      value={value}
+      vertical={vertical}
+      onChange={(e, data) => setValue(String(data.value))}
```
https://github.com/stardust-ui/react/pull/1559/files#r302911394
Do we need this type cast?

bmdalex replied:
yes; reasons:
- I don't think https://github.com/stardust-ui/react/pull/1559/files#r302911394 is the correct fix; the value can be number as well
- `useStringKnob` requires `string` and `value` coming from `Slider` has type `string | number`

layershifter replied:
```
      const value = _.get(e, 'target.value')
      _.invoke(this.props, 'onChange', e, { ...this.props, value })
```

According to these lines it can't be `string | number`:

```js
const e:  React.ChangeEvent<HTMLInputElement> = {}
e.target.value // is `string`
```

See https://github.com/stardust-ui/react/pull/1559#discussion_r302911394

bmdalex replied:
well that was a good argument 😄 applied suggestion, we should find a way to generalize this. thanks, resolving

layershifter commented on docs/src/examples/components/Slider/Usage/SliderExampleAction.shorthand.tsx:
```diff
@@ -0,0 +1,72 @@
+
+interface SliderAction {
+  type: 'toggle_mute' | 'change_value'
+  value?: React.ReactText
```
https://github.com/stardust-ui/react/pull/1559/files#r302911394

bmdalex replied:
done, thanks

layershifter commented on docs/src/examples/components/Slider/Usage/SliderExampleAction.shorthand.tsx:
```diff
@@ -0,0 +1,72 @@
+  })
+
+  const handeChange = React.useCallback(
+    (e, data) => dispatch({ type: 'change_value', value: Number(data.value) }),
```
A bit strange for me, in `Playground` we used `String()`...

Will https://github.com/stardust-ui/react/pull/1559/files#r302911394 solve this?

bmdalex replied:
removed, not needed

layershifter commented on docs/src/examples/components/Slider/Usage/SliderExampleAction.shorthand.tsx:
```diff
@@ -0,0 +1,72 @@
+  }
+}
+
+const SliderWithAction: React.FunctionComponent<
```
Do we really need this a separate component? It looks that this creates an additional complication with types and `commonProps`...

bmdalex replied:
refactored, thanks

layershifter commented on packages/react/src/components/Slider/Slider.tsx:
```diff
@@ -0,0 +1,276 @@
+  value: 50,
+}
+
+const processInputValues = (
```
Can we do this inside `getDerivedStateFromProps()` (`getAutoControlledStateFromProps()`). In that case we will have more accurate values for accessibility and styles...

bmdalex replied:
does that mean `min`, `max` and `valueAsPercentage` will become part of state? why do we need that?
From accessibility POV it works anyway with both strings and numbers.

Are there any other concerns? It seems strange to add these to the state. @layershifter

layershifter replied:
> does that mean `min`, `max` and `valueAsPercentage` will become part of state?

Yep

> From accessibility POV it works anyway with both strings and numbers.
> Are there any other concerns? It seems strange to add these to the state.

In this case: no.

bmdalex committed:
fixed visual tests

layershifter commented on packages/react/src/themes/base/components/Slider/sliderStyles.ts:
```diff
@@ -0,0 +1,92 @@
+import { SliderVariables } from './sliderVariables'
+import { SliderProps, SliderState } from '../../../../components/Slider/Slider'
+import { ComponentSlotStylesInput, ICSSInJSStyle } from '../../../types'
+import { selectors } from '../../../../styles/selectors'
```
Having this in a separate file is confusing for me...

bmdalex replied:
previously I needed `selectors` in more files. moved them back to  `sliderStyles.ts`, thanks

layershifter commented on packages/react/src/themes/base/components/Slider/sliderStyles.ts:
```diff
@@ -0,0 +1,92 @@
+import { ComponentSlotStylesInput, ICSSInJSStyle } from '../../../types'
+import { selectors } from '../../../../styles/selectors'
+
+const getSliderSlotStyles = (p: SliderProps, v: SliderVariables, slot: 'rail' | 'track') => ({
```
_nit_: We can simplify this if we will have:

```jsx
const getCommonSlotStyles = (p: SliderProps, v: SliderVariables) => ({
  cursor: 'pointer',
  pointerEvents: 'none',
  position: 'absolute',
  border: 0,
  height: v.railHeight,
  marginTop: `calc(${v.height} / 2 - ${v.railHeight} / 2)`,
})
```

And conditions below will go back to matching slots:
```jsx
  ...(slot === 'rail' && { width: '100%', background: v.railColor }),
  ...(slot === 'track' && { background: v.trackColor }),
  ...(p.disabled && slot === 'rail' && { background: v.disabledRailColor }),
  ...(p.disabled && slot === 'track' && { background: v.disabledTrackColor }),
```

bmdalex replied:
I like this, done, thanks

layershifter commented on packages/react/src/themes/base/components/Slider/sliderVariables.ts:
```diff
@@ -0,0 +1,48 @@
+import { SiteVariablesPrepared } from '../../../types'
+
+export interface SliderVariables {
+  height: React.CSSProperties['height']
```
Styles should be framework agnostic and variables, too. So we should not use `React.*` stuff

bmdalex replied:
done, thanks

layershifter commented on packages/react/test/specs/components/Slider/Slider-test.tsx:
```diff
@@ -0,0 +1,50 @@
+  describe('auto-controlled', () => {
+    it('sets slider value from user when the value prop is not set (non-controlled mode)', () => {
+      const sliderComp = mount(<Slider />)
+      const domNode = getInputDomNode(sliderComp)
```
Why we need to capture this from DOM? I believe that `expect(wrapper.find('input').prop('value')).toEqual('30')` should be enough

bmdalex replied:
I copied this from `Input-test.tsx`. Any major concerns with it?

layershifter replied:
It's a strange thing to test DOM instead of React in this case. But you can keep it, probably `Input-tests` should be also refactored...

layershifter commented on packages/react/test/specs/components/Slider/Slider-test.tsx:
```diff
@@ -0,0 +1,50 @@
+import { isConformant, handlesAccessibility } from 'test/specs/commonTests'
+import Slider from 'src/components/Slider/Slider'
+
+const testValue = '30'
```
This value is used two times, it will be more obvious if we will just inline it

bmdalex replied:
agreed, done, thanks

layershifter commented on packages/react/test/specs/components/Slider/Slider-test.tsx:
```diff
@@ -0,0 +1,50 @@
+}
+
+describe('Slider', () => {
+  describe('conformance', () => {
```
```suggestion
```

We don't need a separate `describe()` there

bmdalex replied:
done, thanks

bmdalex commented on packages/react/src/components/Slider/Slider.tsx:
```diff
@@ -0,0 +1,276 @@
+ * @accessibility
+ * Implements [ARIA Slider](https://www.w3.org/TR/wai-aria-practices-1.1/#slider) design pattern.
+ */
+export default withSafeTypeForAs<typeof Slider, SliderProps, 'input'>(Slider)
```
when doing this, I cannot access `Slider.defaultProps` outside of this file, @kuzhelov any idea why?

![Screenshot 2019-07-12 at 13 38 58](https://user-images.githubusercontent.com/5442794/61125564-6ce73c80-a4aa-11e9-9d26-93832cc68a9f.png)

bmdalex committed:
addressed PR comments

bmdalex committed:
reverted input changes

bmdalex committed:
moved selectors to styling file

bmdalex committed:
some changes in test files

bmdalex committed:
improved styles for onChange in Slider

mnajdova commented on docs/src/examples/components/Slider/Types/SliderExample.shorthand.steps.ts:
```diff
@@ -0,0 +1,32 @@
+
+const focusSliderStep: ScreenerStep = (builder, keys) => builder.keys('body', keys.tab)
+
+const config: ScreenerTestsConfig = {
```
Let's make sure that we capture this for the rtl example as well

bmdalex replied:
good idea, done

bmdalex committed:
added navigation tests for slider in rtl mode

layershifter reviewed (approved):
👍
